feat(tree): Include list of properties that changed in nodeChanged event

CraigMacomber opened the pull request:
## Description

 Updated version of https://github.com/microsoft/FluidFramework/pull/22043 .

See changeset for details.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

alexvy86 committed:
WIP - test works

alexvy86 committed:
Leaner test

alexvy86 committed:
Use readonly set and remove only from test

alexvy86 committed:
Expose changed properties in simple-tree

alexvy86 committed:
Fix test at simple-tree layer

alexvy86 committed:
Formatting and API reports

alexvy86 committed:
Write simple-tree test in terms of view operations, not editor operations

alexvy86 committed:
Fluid framework API reports

alexvy86 committed:
Test map and array nodes

alexvy86 committed:
Add note

alexvy86 committed:
Formatting

alexvy86 committed:
Merge branch 'main' into node-changed-details

alexvy86 committed:
Return forked checkout object

alexvy86 committed:
Translate to view keys

alexvy86 committed:
Fix circular dependency

alexvy86 committed:
Changeset

alexvy86 committed:
PR feedback

alexvy86 committed:
Move cherk for array case to treeNodeKernel

alexvy86 committed:
Remove unused import

alexvy86 committed:
Move assert to better place

alexvy86 committed:
Update fluid-framework API reports

CraigMacomber committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into node-changed-details

CraigMacomber committed:
Avoid exporting SchematizingSimpleTreeView from simple-tree just to use in simple tree test utils.

CraigMacomber committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into node-change

CraigMacomber committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into node-change

CraigMacomber committed:
specialize for object and map nodes

CraigMacomber committed:
Move event implementation, fix cycle.

CraigMacomber committed:
Fix cyclic deps and add node kind based typing

CraigMacomber commented on packages/dds/tree/src/simple-tree/api/treeNodeApi.ts:
```diff
@@ -167,7 +168,44 @@ export const treeNodeApi: TreeNodeApi = {
+		const kernel = getKernel(node);
+		if (kernel.hydrated === undefined) {
+			throw new UsageError(
+				"Subscribing to events for a node which is Unhydrated is unsupported",
```
Looks like we currently have support for subscribing to unhydrated nodes, which fail to give events while unhydrated, but then start reporting when hydrated.

This seems like both a feature and an undocumented bug. We probably need to preserve that feature, so a different implementation strategy is needed here.

I think I may give the node kernel a hydrated event, or a after hydrate method or something like that.

CraigMacomber replied:
Fixed.

CraigMacomber committed:
Fix unhydrated events

CraigMacomber committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into node-change

CraigMacomber committed:
Update chanegset

Josmithr commented on .changeset/wet-hoops-drop.md:
```diff
@@ -0,0 +1,17 @@
+
+For object nodes, the list of properties uses the property identifiers defined in the schema, and not the persisted
+identifiers (or "stored keys") that can be provided through `FieldProps` when defining a schema.
+See the documentation for `FieldProps` for more details about the distinction between "view keys" and "stored keys".
```
Should we use "property keys" here instead of "view keys"? I thought we preferred the former terminology.

Josmithr commented on packages/dds/tree/src/core/tree/anchorSet.ts:
```diff
@@ -127,7 +127,10 @@ export interface AnchorEvents {
 	 * Compare to {@link AnchorEvents.childrenChanged} which is emitted in the middle of the batch/delta-visit.
 	 */
-	childrenChangedAfterBatch(anchor: AnchorNode): void;
+	childrenChangedAfterBatch(arg: {
```
I realize these eventing methods aren't part of the public API, so this probably doesn't matter too much, but (for consistency) should we go ahead and update all of these methods to take a property bag containing the anchor?

Josmithr commented on packages/dds/tree/src/core/tree/anchorSet.ts:
```diff
@@ -769,10 +778,12 @@ export class AnchorSet implements Listenable<AnchorSetRootEvents>, AnchorLocator
 			notifyChildrenChanged(): void {
 				this.maybeWithNode(
 					(p) => {
+						assert(this.parentField !== undefined, "Must be in a field");
```
Nit: a bit more detail in the error message would probably be useful for future debugging.

Josmithr commented on packages/dds/tree/src/simple-tree/core/treeNodeKernel.ts:
```diff
@@ -60,12 +66,48 @@ export function tryGetTreeNodeSchema(value: unknown): undefined | TreeNodeSchema
+	 * Generation number which is incremented any time we have an edit on the node.
+	 * Used during iteration to make sure there has been no edits that were concurrently made.
+	 */
+	public generationNumber: number = 0;
```
Should mutations to this be more controlled? E.g., make the setter private and expose an `increment` method - otherwise consumers could potentially make invalid changes to this.

CraigMacomber replied:
Personally for very internal implementation detail code like this (not even exported out of simple-tree) I don't think its worth the extra boiler plate, runtime overhead and bundle size to make that change. If someday we want a bit more encapsulation here we can make a TreeNodeKernel interface so we can have that improved safety without the runtime or bundle size overhead.

A decent portion of fluid runtime bundle size is wrappers around class properties to prevent external writing (they are a few percent of one of our largest files, see packages/runtime/container-runtime/src/containerRuntime.ts): I don't think that pattern is worth the cost here.

Josmithr commented on packages/dds/tree/src/simple-tree/core/treeNodeKernel.ts:
```diff
@@ -60,12 +66,48 @@ export function tryGetTreeNodeSchema(value: unknown): undefined | TreeNodeSchema
+
+	/**
+	 * Events registered before hydration.
+	 * @remarks
```
TODO?

Josmithr commented on packages/dds/tree/src/simple-tree/core/treeNodeKernel.ts:
```diff
@@ -80,37 +122,50 @@ export class TreeNodeKernel implements Listenable<TreeChangeEvents> {
-		});
-
-		const offAfterDestroy = anchorNode.on("afterDestroy", () => this.dispose());
+		assert(!this.disposed, "cannot use a disposed node");
```
Nit
```suggestion
		assert(!this.disposed, "cannot hydrate a disposed node");
```

Josmithr commented on packages/dds/tree/src/simple-tree/core/treeNodeKernel.ts:
```diff
@@ -80,37 +122,50 @@ export class TreeNodeKernel implements Listenable<TreeChangeEvents> {
-			},
+			offAnchorNode: new Set([
+				anchorNode.on("afterDestroy", () => this.dispose()),
+				// TODO: this should be triggered on change even for unhydrated nodes.
```
Does this need to be done now? What problems are caused by the current situation?

CraigMacomber replied:
you will not get any errors if you iterate unhydrated nodes while mutating them. I think thats thew only thing we currently use generation numbers for.

CraigMacomber replied:
note that this is a preexisting issue unrelated to this PR which is simply being documented.

Josmithr commented on packages/dds/tree/src/simple-tree/core/treeNodeKernel.ts:
```diff
@@ -80,37 +122,50 @@ export class TreeNodeKernel implements Listenable<TreeChangeEvents> {
-		this.#hydrated?.offAnchorNode?.();
-		this.#hydrated = undefined;
+		// If needed, register forwarding emitters for events from before hydration
+		if (this.#preHydrationEvents !== undefined) {
```
Nit: early return on `this.#preHydrationEvents === undefined` would reduce nesting some here.

CraigMacomber replied:
I prefer this as is, since if someone needs to add more logic to this function in the future, this way they can just add it at the top level in what ever order it makes sense, whereas if we make your proposed change, if they add it to the end of the function, it won't run if there were prehydration events, which is almost certainly unrelated to what ever logic they might want to add.

Josmithr commented on packages/dds/tree/src/simple-tree/core/types.ts:
```diff
@@ -24,9 +24,27 @@ import { isFlexTreeNode, type FlexTreeNode } from "../../feature-libraries/index
 
+/**
+ * Data included for {@link TreeChangeEvents.nodeChanged}.
+ * @public
```
Should this be `@sealed`?

Josmithr commented on packages/dds/tree/src/simple-tree/core/types.ts:
```diff
@@ -24,9 +24,27 @@ import { isFlexTreeNode, type FlexTreeNode } from "../../feature-libraries/index
+	 * @remarks
+	 * This only includes changes to the node itself (which would trigger {@link TreeChangeEvents.nodeChanged}).
+	 */
+	readonly changedProperties?: ReadonlySet<string>;
```
Presumably `undefined` here is equivalent to an empty set? Might be worth calling that out explicitly.

CraigMacomber replied:
no, not equivalent. I'll update the docs.

CraigMacomber replied:
updated

Josmithr commented on packages/dds/tree/src/simple-tree/core/types.ts:
```diff
@@ -45,17 +63,15 @@ export type Unhydrated<T> = T;
  * @sealed @public
  */
-export interface TreeChangeEvents {
+export interface TreeChangeEvents<TNode = TreeNode> {
```
Sanity check: should this be `TNode extends TreeNode = TreeNode`?

Josmithr commented on packages/dds/tree/src/simple-tree/core/types.ts:
```diff
@@ -45,17 +63,15 @@ export type Unhydrated<T> = T;
 	 *
-	 * - For an object node, when the value of one of its properties changes (i.e., the property's value is set
-	 * to something else, including `undefined`).
+	 * - Object nodes define a change as the value of one of its properties changing (i.e., the property's value is set, including when set to undefined).
```
Nit (I initially had a hard time parsing this, but understood after re-reading. Feel free to take or leave this suggestion):
```suggestion
	 * - Object nodes define a change as being when the value of one of its properties changes (i.e., the property's value is set, including when set to `undefined`).
```

Also, should it say "one or more of its properties changes"?

Josmithr commented on packages/dds/tree/src/simple-tree/core/types.ts:
```diff
@@ -45,17 +63,15 @@ export type Unhydrated<T> = T;
+	 * - Object nodes define a change as the value of one of its properties changing (i.e., the property's value is set, including when set to undefined).
 	 *
-	 * - For an array node, when an element is added, removed, or moved.
+	 * - Array node define a change as when an element is added, removed, moved or replaced.
```
```suggestion
	 * - Array nodes define a change as when an element is added, removed, moved or replaced.
```

Josmithr commented on packages/dds/tree/src/simple-tree/core/types.ts:
```diff
@@ -70,15 +86,23 @@ export interface TreeChangeEvents {
 	 * node, or when the node has to be updated due to resolution of a merge conflict
 	 * (for example a previously applied local change might be undone, then reapplied differently or not at all).
+	 *
+	 * TODO: defined and document event ordering (ex: bottom up, with nodeChanged before treeCHange on each level).
```
```suggestion
	 * TODO: define and document event ordering (ex: bottom up, with nodeChanged before treeCHange on each level).
```

alexvy86 replied:
Also `treeCHange => treeChange` (lowercase H).

Josmithr commented on packages/dds/tree/src/simple-tree/core/types.ts:
```diff
@@ -99,6 +123,8 @@ export interface TreeChangeEvents {
 	treeChanged(): void;
 }
 
+export type IsListener2<TListener> = TListener extends (...args: any[]) => void ? true : false;
```
Is this used?

Josmithr commented on packages/dds/tree/src/test/simple-tree/api/treeApi.spec.ts:
```diff
@@ -768,5 +768,114 @@ describe("treeNodeApi", () => {
+				eventLog.push(changedProperties),
+			);
+
+			const { forkView, forkCheckout } = getViewForForkedBranch(view);
```
For my knowledge: is this the recommended pattern for bundling multiple tree operations into a single transaction?

Josmithr replied:
Ah, I see this is a newly added function and not exported. Do we have a recommended pattern for external users?

CraigMacomber replied:
if you want to group operations, use transactions.

This code is using branches, presumably to simulate if the ops were applied by another client, but without requiring a runtime, multiple tree instances etc. This is a common pattern in our tests since branches and multiple clients hit the same code paths, but using branches involves far fewer dependencies.

This same scenario however would work fine without the branch since local edits also trigger events, so its unclear why its making a branch.

Josmithr commented on packages/dds/tree/src/test/simple-tree/unhydratedNode.spec.ts:
```diff
@@ -177,6 +177,24 @@ describe("Unhydrated nodes", () => {
+		hydrate(TestLeaf, leafObject);
+		leafObject.value = "new value";
+		// Assert that the event fired
+		// TODO: Eventually the order of events should be documents, and an approach like this can test that they are ordered as documented.
```
```suggestion
		// TODO: Eventually the order of events should be documents, and an approach like this can test that they are ordered as documented.
```
```suggestion
		// TODO: Eventually the order of events should be documented, and an approach like this can test that they are ordered as documented.
```

Josmithr commented on packages/dds/tree/src/test/tree/anchorSet.spec.ts:
```diff
@@ -492,6 +492,49 @@ describe("AnchorSet", () => {
+		node0.on("childrenChangedAfterBatch", ({ anchor, changedFields }) => {
+			// This is the main validation of this test
+			assert.deepEqual(changedFields, expectedChangedFields);
+			listenerFired = true;
```
Nit: should we ensure that the listener is only called _once_?

alexvy86 replied:
In fact, by applying the pattern Craig suggested to me before and is used in some of the other new tests. Keep a log where all event payloads are pushed when the event fires, and in the end check that both the correct number of entries are present and they have the expected structure.

alexvy86 commented on packages/dds/tree/src/simple-tree/core/types.ts:
```diff
@@ -70,15 +86,23 @@ export interface TreeChangeEvents {
 	 */
-	nodeChanged(): void;
+	nodeChanged(
+		data: NodeChangedData &
```
If I understand this correctly, for array nodes the API will still say there is an optional `<listenerArg>.changedProperties`, but at runtime it's always undefined? So there's no way to completely hide it from the listener's signature in that case?

CraigMacomber replied:
Removing it when passing in a node known to be an array node is possible, but that adds additional typing complexity to the signature and I don't think offers any value.

If someone knows their node is an array, seeing the changedProperties value and being able to read its docs doesn't seem like a bad thing. I don't imagine anyone will write code which would be able to be shorter or simpler if we hide that property from the array case. This is not true for making it never undefined for object and map nodes as such code would otherwise need to check for undefined before reading the set. Array specific code on the other hand probably will never need to check for it being a set before reading from the undefined (as there is nothing to read from an undefined value))

alexvy86 commented on packages/dds/tree/src/test/simple-tree/unhydratedNode.spec.ts:
```diff
@@ -177,6 +177,24 @@ describe("Unhydrated nodes", () => {
 		assert.equal(Tree.status(object), TreeStatus.New);
 	});
 
+	it("preserve events after hydration - minimal", () => {
```
```suggestion
	it("preserve event subscriptions during hydration - minimal", () => {
```

alexvy86 commented on packages/dds/tree/src/test/simple-tree/unhydratedNode.spec.ts:
```diff
@@ -177,6 +177,24 @@ describe("Unhydrated nodes", () => {
 
+	it("preserve events after hydration - minimal", () => {
+		const log: unknown[] = [];
+		const leafObject = new TestLeaf({ value: "value" });
```
Just a note for clarity/posterity: it initially confused me to see events being subscribed to on (what I thought was) a leaf. Then I realized `TestLeaf extends schemaFactory.object(...)` so it's really an object node. Just noting my confusion; maybe we want to use a different name instead of `TestLeaf` for scenarios like this.

alexvy86 commented on packages/dds/tree/src/simple-tree/core/treeNodeKernel.ts:
```diff
@@ -60,12 +66,48 @@ export function tryGetTreeNodeSchema(value: unknown): undefined | TreeNodeSchema
+	 * This design avoids allocating `#preHydrationEvents` if unneeded.
+	 *
+	 * This design also avoids extra forwarding overhead for events from anchorNode and also
+	 * avoids registering for events that the are unneeded.
```
I think?
```suggestion
	 * avoids registering for events that are unneeded.
```

alexvy86 commented on packages/dds/tree/src/simple-tree/core/treeNodeKernel.ts:
```diff
@@ -127,15 +182,19 @@ export class TreeNodeKernel implements Listenable<TreeChangeEvents> {
 
 	public dispose(): void {
-		this.dehydrate();
+		this.disposed = true;
```
Would it make sense to also set `this.#hydrated = undefined` like the current code does, to avoid potential future bugs if someone forgets to check `this.disposed` before trying to use `this.#hydrated`?

CraigMacomber replied:
I don't like making the node look like it never hydrated when we dispose of it. My understanding is that hydration is suppose to be a one time thing and there is logic that depend on nodes never unhydrating.

If something fails to check disposed, I'd much rather have a node show up as hydrated than unhydrated if its a disposed hydrated node.

The old code was really confusing about this: I couldn't tell if it was trying to support an Unhydration actions. I don't think we want to support that, so I tried to make it clear we never do it.

alexvy86 reviewed (commented):
Thanks for taking over this change!

CraigMacomber commented on packages/dds/tree/src/core/tree/anchorSet.ts:
```diff
@@ -177,11 +180,6 @@ export interface AnchorEvents {
-	/**
-	 * Value on this node is changing.
-	 */
-	valueChanging(anchor: AnchorNode, value: Value): void;
```
Yes, Also because we do not support modifying values of nodes so it never could be used.

CraigMacomber committed:
Apply suggestions from code review

Co-authored-by: Alex Villarreal <716334+alexvy86@users.noreply.github.com>
Co-authored-by: Joshua Smithrud <54606601+Josmithr@users.noreply.github.com>

CraigMacomber committed:
Better docs

CraigMacomber committed:
More comment fixes

CraigMacomber commented:
A usability issue with this APi has been reported:

Lack of strong typing of the keys least to string names for keys in code easily getting out of date (or having typos) with no compile errors.

There are two ways this could be improved:

- strong types on the keys in the set (ex: derive them from the object node or nodes provided)
- a way to refer to property key strings that is strongly typed to the object they are on and refactors properly. For example `Foo.fields.bar.key` possible leveraging a "FieldSchemaContextual" subclass which has the key as s strongly typed property.

CraigMacomber committed:
Refer to property keys not view keys

CraigMacomber committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into node-change

CraigMacomber committed:
Update APi reports

CraigMacomber committed:
Make new event API beta

CraigMacomber committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into node-change

CraigMacomber committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into node-change

CraigMacomber committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into node-change

CraigMacomber committed:
Add strongly typed keys

CraigMacomber committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into node-change

CraigMacomber commented:
> A usability issue with this APi has been reported:
> 
> Lack of strong typing of the keys least to string names for keys in code easily getting out of date (or having typos) with no compile errors.
> 
> There are two ways this could be improved:
> 
> * strong types on the keys in the set (ex: derive them from the object node or nodes provided)
> * a way to refer to property key strings that is strongly typed to the object they are on and refactors properly. For example `Foo.fields.bar.key` possible leveraging a "FieldSchemaContextual" subclass which has the key as s strongly typed property.

Fixed. Also moved API to beta incase there are more such issues.

Josmithr commented on packages/dds/tree/api-report/tree.beta.api.md:
```diff
@@ -401,12 +406,22 @@ interface TreeArrayNodeBase<out T, in TNew, in TMoveFrom> extends ReadonlyArray<
 // @public @sealed
 export interface TreeChangeEvents {
-    nodeChanged(): void;
+    nodeChanged(unstable?: unknown): void;
```
What is the purpose of this change? Is this required? Or just a signal that we plan on adding data to the event in the future?

CraigMacomber replied:
I forgot to document that... will do

CraigMacomber replied:
Done.

CraigMacomber committed:
document unstable

CraigMacomber committed:
Merge branch 'main' into node-change
